### PR TITLE
Fix #6951: Unexpected error during scheme creation.

### LINKED ISSUE #6951: Unexpected error during scheme creation: Anomaly "Uncaught exception Type_errors.TypeError(_, _)."
#### Version

8.8

#### Description of the problem

```
Record float2 : Set := Float2 { Fnum : unit }.
Scheme Equality for float2.

Unexpected error during scheme creation:
Anomaly "Uncaught exception Type_errors.TypeError(_, _)."
```

This is a regression with respect to Coq 8.7.

### PR DESCRIPTION
**Pull request by ppedrot**
When creating a scheme for bifinite inductive types, we do not create a fixpoint.

### CODE REVIEW

**Comment by maximedenes** [comment]
The fix is a bit dirty, but given that this was a serious regression...

@mattam82 can you have a look?

**Comment by ppedrot** [comment]
By the way, there is something wrong in the check in the kernel. The generated `float2_beq` still contains a fixpoint, but its recursive argument is bifinite, and the kernel accepts that.

File: vernac/auto_ind_decl.ml [hunk 1]
```diff
@@ -319,9 +319,17 @@ let build_beq_scheme mode kn =
...
+        | CoFinite ->
 	  raise NoDecidabilityCoInductive;
-        let fix = mkFix (((Array.make nb_ind 0),i),(names,types,cores)) in
+        | Finite ->
+          mkFix (((Array.make nb_ind 0),i),(names,types,cores))
+        | BiFinite ->
+          (** If the inductive type is not recursive, the fixpoint is not
+              used, so let's replace it with garbage *)
+          let subst = List.init nb_ind (fun _ -> mkProp) in
+          Vars.substl subst cores.(i)
```
Comment (SkySkimmer) [inline vernac/auto_ind_decl.ml]: I always get lost with debruijns but isn't this some kind of lift?

Comment (ppedrot) [reply vernac/auto_ind_decl.ml]: I don't think so, at least theoretically. Maybe the implementation of lift works for negative lifts, but there is no reason it behaves properly for strenghtening.

**Review by mattam82 — APPROVED** [review]

**Comment by mattam82** [comment]
Wait, reading @ppedrot 's last comment, I don't understand: how can the code avoid the code path checking that the fix is on a Finite inductive? Is the check in #407 really that weak?

**Comment by mattam82** [comment]
I don't think tactics generate fixpoints, except for `fix` of course. On the command side, it leaves `Lemma .. with ..` I think, maybe others.

**Comment by maximedenes** [comment]
what about `decide equality`, other `Scheme` variants, etc?

**Comment by mattam82** [comment]
I don't have all the Schemes in mind, but I think the elimination principles are properly restricted and Scheme Equality is the only other one, but I might be wrong. Right, I forgot about `decide equality`.

**Comment by maximedenes** [comment]
@ppedrot I see some failures. Are they related?

**Comment by maximedenes** [comment]
This is now too late anyway (sorry), let's postpone to 8.8.1.

**Comment by ppedrot** [comment]
Equations and elpi are broken for unrelated reasons. I don't know about VST, but I'll have a look.

**Comment by ppedrot** [comment]
I cannot reproduce the VST failure either.

**Comment by maximedenes** [comment]
@mattam82 ping

**Comment by maximedenes** [comment]
The ping is about getting this merged, it's already a pitty that it missed 8.8.0.

**Comment by ppedrot** [comment]
@maximedenes I currently have physical access to @mattam82 and he swore he was about to merge a lot of PRs... I can be slightly more persuasive if needed!

**Comment by maximedenes** [comment]
This has been delayed for too long, I'll merge as the secondary maintainer.

**Pull request merged**